bzlmod: Add dedicated documentation (#3519)
bazelbuild/rules_go: docs/go/core/bzlmod.md

# Go with Bzlmod

This document describes how to use rules_go and Gazelle with Bazel's new external dependency subsystem [Bzlmod](https://bazel.build/external/overview#bzlmod), which is meant to replace `WORKSPACE` files eventually.

> Usages of rules_go and Gazelle in `BUILD` files are not affected by this, refer to the existing documentation on rules and configuration options for them.

Reviewer: ```suggestion
Usages of rules_go and Gazelle in `BUILD` files are not affected by this; refer to the existing documentation on rules and configuration options for them.
```

## Setup

Add the following lines to your `MODULE.bazel` file:

```starlark
bazel_dep(name = "rules_go", version = "0.39.1")
bazel_dep(name = "gazelle", version = "0.29.0")
```

The latest versions are always listed on https://registry.bazel.build/.

If you have WORKSPACE dependencies that reference rules_go and/or Gazelle, you can still use the legacy repository names for the two repositories:

```starlark
bazel_dep(name = "rules_go", version = "0.39.1", repo_name = "io_bazel_rules_go")
bazel_dep(name = "gazelle", version = "0.29.0", repo_name = "bazel_gazelle")
```

## Registering Go SDKs

rules_go automatically downloads and registers a recent Go SDK, so unless a particular version is required, no manual steps are required.

To register a particular version of the Go SDK, use the `go_sdk` module extension:

```starlark
go_sdk = use_extension("@rules_go//go:extensions.bzl", "go_sdk")

# Download an SDK for the host OS & architecture.
go_sdk.download(version = "1.20.3")

> # Download an SDK for a fixed OS/architecture, e.g. for remote execution.

Reviewer: Assuming that people usually only want one of these three options:
```suggestion
# Alternately, download an SDK for a fixed OS/architecture, e.g. for remote execution.
```

go_sdk.download(
    version = "1.20.3",
    goarch = "amd64",
    goos = "linux",
)

> # Register the Go SDK installed on the host.

Reviewer: Assuming that people usually only want one of these three options:
```suggestion
# Alternately, register the Go SDK installed on the host.
```
I see that you mention `go_cross_binary` below, so maybe people _will_ want to register several SDKs together. If they do, how will they identify each of them later? Should we show use of the "name" attribute here?

Author: `go_cross_binary` can change the platform and the required SDK version. We generally want people to specify what they want and receive a matching SDK without caring about which particular one they want. If they really want to pin a particular one, they should use `name` and `--extra_toolchains`, but that should be very rare.

go_sdk.host()
```

> Multiple Go SDKs can be registered and selected on a per-target basis using [`go_cross_binary`](rules.md#go_cross_binary).

Reviewer: ```suggestion
You can register multiple Go SDKs and select which one to use on a per-target basis using [`go_cross_binary`](rules.md#go_cross_binary).
```

### Not yet supported

* `go_local_sdk`
* `go_wrap_sdk`
* nogo

## Generating BUILD files

Add the following to your top-level BUILD file, including your Go module's path in a [Gazelle directive](https://github.com/bazelbuild/bazel-gazelle#directives):

```starlark
load("@gazelle//:defs.bzl", "gazelle")

# gazelle:prefix github.com/example/project
gazelle(name = "gazelle")
```

Then, use `bazel run //:gazelle` to (re-)generate BUILD files.

## External dependencies

External Go dependencies are managed by the `go_deps` module extension provided by Gazelle.
`go_deps` performs [Minimal Version Selection](https://go.dev/ref/mod#minimal-version-selection) on all transitive Go dependencies of all Bazel modules, so compared to the old WORKSPACE setup, every Bazel module only needs to declare its own Go dependencies.
For every major version of a Go module, there will only ever be a single version in the entire build, just as in regular Go module builds.

### Specifying external dependencies

Even though this is not a strict requirement, for interoperability with Go tooling that isn't Bazel-aware, it is recommended to manage Go dependencies via `go.mod`.

> The `go_deps` extension can directly parse this file, so external tooling such as `gazelle update-repos` is no longer needed.

Reviewer: ```suggestion
The `go_deps` extension parses this file directly, so external tooling such as `gazelle update-repos` is no longer needed.
```

Register the `go.mod` file with the `go_deps` extension as follows:

```starlark
go_deps = use_extension("@gazelle//:extensions.bzl", "go_deps")
go_deps.from_file(go_mod = "//:go.mod")

# All *direct* Go dependencies of the module have to be listed explicitly.
use_repo(
    go_deps,
    "com_github_gogo_protobuf",
    "com_github_golang_mock",
    "com_github_golang_protobuf",
    "org_golang_x_net",
)
```

> There is [ongoing work](https://github.com/bazelbuild/bazel/issues/17908) targeted for Bazel 6.2.0 to automate the generation of the `use_repo` statement.

Reviewer: I'm not sure whether bazelbuild/bazel-gazelle#1423 will remain relevant if we eliminate the need to mention these repository names here, but there was the idea to mention Go module paths instead.

Author: That is still on my radar, but I first want to see how far Bazel labels can get us here while remaining usable.

Alternatively, you can specify a module extension tag to add an individual dependency.

> This can for example be useful for dependencies of generated code that `go mod tidy` would remove (there is [ongoing work](https://github.com/bazelbuild/bazel-gazelle/pull/1495) to provide a Bazel-aware version of `tidy`).

Reviewer: ```suggestion
This can be useful for dependencies of generated code that `go mod tidy` would remove. (There is [ongoing work](https://github.com/bazelbuild/bazel-gazelle/pull/1495) to provide a Bazel-aware version of `tidy`.)
```

    version = "v1.50.0",
)
```

### Managing `go.mod`

An initial `go.mod` file can be created via

```sh
bazel run @rules_go//go mod init github.com/example/project
```

A dependency can be added via

```sh
bazel run @rules_go//go get golang.org/x/text@v0.3.2
```

### Overrides

The root module can override certain aspects of the dependency resolution performed by the `go_deps` extension.

#### `replace`

[`replace` directives](https://go.dev/ref/mod#go-mod-file-replace) in `go.mod` can be used to replace particular versions of dependencies with other versions or entirely different modules.
At the moment the only supported form is:

```
replace(
    golang.org/x/net v1.2.3 => example.com/fork/net v1.4.5
)
```

#### Gazelle directives

Some external Go modules may require tweaking how Gazelle generates BUILD files for them via [Gazelle directives](https://github.com/bazelbuild/bazel-gazelle#directives).
The `go_deps` extension provides a dedicated `go_deps.gazelle_override` tag for this purpose:

```starlark
go_deps.gazelle_override(
    directives = [
        "gazelle:go_naming_convention go_default_library",
    ],
    path = "github.com/stretchr/testify",
)
```

> If you need to use a `gazelle_override` to get a public Go module to build with Bazel, consider contributing the directives to the [public registry for Gazelle directives](internal/bzlmod/directives.bzl) via a PR.

Reviewer 2: I think the link for `public registry...` needs to begin with a `/` maybe?

Author: It actually has to be an external link since the `bzl` file lives in Gazelle. Fixed, thanks!
